cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -0,0 +1,195 @@
+
+  // The file base name (component after the last directory separator or '/').
+  std::string base_name() const;
+
+  // The size in bytes, if available.  Only regular files are guaranteed
+  // to have a size.
+  int64_t size() const { return size_; }
+  void size(int64_t size) { size_ = size; }
+
+  // The time of last modification, if available.
+  TimePoint mtime() const { return mtime_; }
+  void mtime(TimePoint mtime) { mtime_ = mtime; }
+
+ protected:
+  FileType type_ = FileType::Unknown;
+  std::string path_;
+  int64_t size_ = -1;
+  TimePoint mtime_ = TimePoint(TimePoint::duration(-1));
+};
+
+/// \brief EXPERIMENTAL: stream / file open flags
+struct ARROW_EXPORT OpenFlags {
+  // Whether to transparently compress / decompress.
+  Compression::type compression = Compression::DETECT;
+
+  OpenFlags() {}
+};
+
+/// \brief EXPERIMENTAL: file selector
+struct ARROW_EXPORT Selector {
+  // The directory in which to select files.
+  // If the path exists but doesn't point to a directory, this should be an error.
+  std::string base_dir;
+  // The behavior if `base_dir` doesn't exist in the filesystem.  If false,
+  // an error is returned.  If true, an empty selection is returned.
+  bool allow_non_existent = false;
+  // Whether to recurse into subdirectories.
+  bool recursive = false;
+
+  Selector() {}
+};
+
+/// \brief EXPERIMENTAL: abstract file system API
+class ARROW_EXPORT FileSystem {
+ public:
+  virtual ~FileSystem();
+
+  /// Get statistics for the given target.
+  ///
+  /// Any symlink is automatically dereferenced, recursively.
+  /// A non-existing or unreachable file returns an Ok status and
+  /// has a FileType of value NonExistent.  An error status indicates
+  /// a truly exceptional condition (low-level I/O error, etc.).
+  virtual Status GetTargetStats(const std::string& path, FileStats* out) = 0;
+  /// Same, for many targets at once.
+  virtual Status GetTargetStats(const std::vector<std::string>& paths,
+                                std::vector<FileStats>* out);
+  /// Same, according to a selector.
+  ///
+  /// The selector's base directory will not be part of the results, even if
+  /// it exists.
+  /// If it doesn't exist, see `Selector::allow_non_existent`.
+  virtual Status GetTargetStats(const Selector& select, std::vector<FileStats>* out) = 0;
+
+  /// Create a directory and subdirectories.
+  virtual Status CreateDir(const std::string& path, bool recursive = true) = 0;
+
+  /// Delete a directory and contents.
+  virtual Status DeleteDir(const std::string& path) = 0;
+
+  /// Delete a file.
+  virtual Status DeleteFile(const std::string& path) = 0;
+  /// Delete many files.
+  ///
+  /// The default implementation issues individual delete operations in sequence.
+  virtual Status DeleteFiles(const std::vector<std::string>& paths);
+
+  /// Move / rename a file or directory.
+  ///
+  /// The destination will be replaced if it exists.
+  // XXX separate MoveFile / MoveDir ?
```

Comment:
I guess it's OK to be both for now. In HDFS I wrote `Rename` for both directories and files

https://github.com/apache/arrow/blob/master/cpp/src/arrow/io/hdfs.h#L139

Some filesystems may have to stat the path to determine what it is, where some don't have to, so I think having a single API makes sense

Reply:
Fair enough.
Note that "stat first then choose the API" is vulnerable to race conditions. But for our purposes I'm not sure we care about such issues.

Reply:
Yeah, I think we can leave it to the particular filesystem implementation to decide what is most appropriate to do